python-lib/dku_deeplearning_image/utils.py

```diff
@@ -169,24 +175,50 @@ def format_predictions_output(predictions, classify=False, labels_df=None, limit
         return predictions.tolist()
     formatted_predictions = []
     id_pred = lambda index: labels_df.loc[index][constants.LABEL] if labels_df is not None else str(index)
+    pred_errors = []
     for pred in predictions:
-        formatted_pred = get_ordered_dict(
-            {id_pred(i): float(pred[i]) for i in pred.argsort()[-limit:] if float(pred[i]) >= min_threshold})
-        formatted_predictions.append(formatted_pred)
-    return formatted_predictions
-
-
-def get_predictions(model, batch, classify=False, limit=constants.DEFAULT_PRED_LIMIT, min_threshold=0, labels_df=None):
-    predictions = model.predict(batch)
-    return format_predictions_output(predictions, classify, labels_df, limit, min_threshold)
+        try:
+            formatted_pred = get_ordered_dict(
+                {id_pred(i): float(pred[i]) for i in pred.argsort()[-limit:] if float(pred[i]) >= min_threshold})
+            formatted_predictions.append(formatted_pred)
+            pred_errors.append(0)
+        except Exception:
+            pred_errors.append(1)
+    return {"prediction": formatted_predictions, "error": pred_errors}
 
 
 def get_ordered_dict(predictions):
     return json.dumps(OrderedDict(sorted(predictions.items(), key=(lambda x: -x[1]))))
 
 
-def preprocess_img(img_path, img_shape, preprocessing):
+def apply_preprocess_image(tfds, input_shape, preprocessing, is_b64=False):
+    def _apply_preprocess_image(image_path):
+        return tf.numpy_function(
+            func=lambda x: preprocess_img(x, input_shape, preprocessing, is_b64),
+            inp=[image_path],
+            Tout=tf.float32)
+
+    return tfds.map(map_func=_apply_preprocess_image, num_parallel_calls=constants.AUTOTUNE)
+
+
+def read_images_to_tfds(images_folder, np_images):
+    def retrieve_image_from_folder(image_fn):
+        return tf.numpy_function(
+            func=lambda x: get_cached_file_from_folder(images_folder, x, is_byte=True),
+            inp=[image_fn],
+            Tout=tf.string)
+
+    X_tfds = tf.data.Dataset.from_tensor_slices(np_images.reshape(-1, 1))
+    return X_tfds.interleave(
```

Comment:
why use `interleave` rather than `map` here ?

Reply:
`interleave` is the function made for data extraction (https://www.tensorflow.org/guide/data_performance#parallelizing_data_extraction) and `map` for data transformation. I do not really understand the difference between these both, but I'd rather follow tf recommandations.

Reply:
I think `interleave` usecase is more when each call builds a `Dataset` (one -> many) while `map` builds only element of the resulting dataset (one -> one).

The typical usecase for interleave is when you want to build a text Dataset out of multiple files. Each file reading will yield one dataset (where each line is an element) that you want to merge into a single big dataset for the training. 

Here you just want to map one file to one image, so I think using `map` is clearer. What do you think ?